balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,626 @@
+		b.addressIndex.seekTo(sd.addr)
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, reset index and re-connect when
+	// prompted.
+	if b.state == connectivity.Ready && state.ConnectivityState == connectivity.Idle {
+		// Once a transport fails, we enter idle and start from the first address
+		// when the picker is used.
+		b.state = connectivity.Idle
+		b.addressIndex.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			// We can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If we're in TRANSIENT_FAILURE, we stay in TRANSIENT_FAILURE until
+			// we're READY. See A62.
+			// If we're already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = state.ConnectionError
+			// Since we're re-using common subconns while handling resolver updates,
+			// we could receive an out of turn TRANSIENT_FAILURE from a pass
+			// over the previous address list. We ignore such updates.
+			curAddr := b.addressIndex.currentAddress()
+			if activeSD, found := b.subConns.Get(curAddr); !found || activeSD != sd {
+				return
+			}
+			if b.addressIndex.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass(state.ConnectionError)
+		case connectivity.Idle:
+			// A subconn can transition from CONNECTING directly to IDLE when
+			// a transport is successfully created, but the connection fails before
+			// the subconn can send the notification for READY. We treat this
+			// as a successful connection and transition to IDLE.
+			curAddr := b.addressIndex.currentAddress()
+			if activeSD, found := b.subConns.Get(curAddr); !found || activeSD != sd {
+				return
+			}
+			b.state = connectivity.Idle
+			b.addressIndex.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.Idle,
+				Picker:            &idlePicker{exitIdle: b.ExitIdle},
+			})
+		}
+		return
+	}
+
+	// We have finished the first pass, keep re-connecting failing subconns.
+	switch state.ConnectivityState {
+	case connectivity.TransientFailure:
+		sd.lastErr = state.ConnectionError
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: state.ConnectionError},
+		})
+		// We don't need to request re-resolution since the subconn already does
+		// that before reporting TRANSIENT_FAILURE.
```

Comment:
Nit: Add a TODO with a link to the issue that you created about moving re-resolution to the pick_first policy (and out of the subchannel).